### ISSUE TITLE
API for `Contig` class and `Species.get_contig()`.

### ISSUE THREAD

**Issue by grahamgower** [popsim-consortium/stdpopsim#670]
We've recently had a number of ideas that require or suggest changes to the `Contig` class or the `get_contig()` method. I've opened this issue so that we can think about the API more concretely. It looks like we're going to end up with lots of new parameters to `get_contig()`, so I'll start with describing some of the semantics for various use cases (these naturally extend to hypothetical CLI params too).

Currently possible:
* `get_contig("chr1")`. Simulate chromosome 1. Use a uniform recombination rate matching chr1.
* `get_contig("chr1", genetic_map="foo")`. Simulate chromosome 1 using the "foo" genetic map.
* `get_contig("chr1", length_multiplier=0.01)`. I don't care what it is, just simulate something. And be quick about it! Use a uniform recombination rate matching chr1.

Proposed:
* `get_contig(length=1e6)`. Simulate 1 Mb. Use a uniform recombination rate matching the genome-wide average. #668/#669
* `get_contig("chr1", genetic_map="foo", mask=unmappable_intervals)`. Simulate chromosome 1 using the "foo" genetic map, but mask out the regions in "unmappable_intervals" (this could be some array data structure, or maybe a .bed filename?). #664
* `get_contig("chr1", start=32e6, length=1e6, genetic_map="foo")`. Simulate 1Mb. Use the section of the recombination map from chr1:32Mb-33Mb. #401/#402
* `get_contig(length=1e6, genetic_map="foo")`. Simulate 1 Mb. Use a 1Mb chunk of the "foo" genetic map, sampled uniformly at random. #401/#402
* `get_contig(length=1e6, genetic_map="foo", mask=unmappable_intervals)`. Simulate 1 Mb. Use a 1Mb chunk of the "foo" genetic map, sampled uniformly at random from intervals not in the "unmappable_intervals". (proposed interaction of #668 and #401).

Invalid configurations:
* `length_multiplier` with any of: `genetic_map`,`mask`, or `length`.

---
In #644, @mufernando is adding new `mutation_types` and `genomic_element_types` attributes to the Contig class in order to handle distinct mutation types/rates on distinct genomic annotations. This should naturally combine with many of the above use cases.
* `get_contig(length=1e6)` with `some_BGS_model()`. Simulate 1 Mb, with 1Mb of annotations drawn uniformly at random from the annotations used by the given BGS model.
* `get_contig(length=1e6, genetic_map="foo")` with `some_BGS_model()`. Simulate 1 Mb. Sample a 1Mb genomic region uniformly from the genome, and use this region for the recombination map and the annotations relevant to the given BGS model.
* `get_contig("chr1", start=32e6, length=1e6, genetic_map="foo")` with `some_BGS_model()`. Simulate 1Mb from chr1:32Mb-33Mb. Use the genetic map and relevant annotations for that interval.

It seems plausible that in practise we'll simulate a BGS model like this:
```
species = stdpopsim.get_species(...)
mutation_types, genomic_element_types = some_BGS_model()
contig = species.get_contig(..., mutation_types=mutation_types, genomic_element_types=genomic_element_types)
```
And it should be invalid to combine these with `length_multiplier`.

[EDIT: removed `+` symbol, to avoid suggesting we overload the plus operator]

**Comment by jeromekelleher** [comment]
Sounds good, thanks for bringing this together @grahamgower. A couple of things:

- I'm happy to drop ``length_multiplier`` entirely. The ``get_contig(length=1e6)`` form satisfies that I want to do most often, I think.
- Do we want to support positive and negative masking (#644) ? It seems likely we'll need both, so maybe we also have a keyword ``keep`` which takes a list of intervals to keep? The names ``keep`` and ``mask`` seem logical enough to me for this (but just throwing out ideas). 
- Overloading `+` makes me very nervous. Why not make the bgs_model a keyword arg for ``get_contig``?

**Comment by grahamgower** [comment]
> * I'm happy to drop `length_multiplier` entirely

I think we can formally deprecate it (with a warning) and not remove it.

> Do we want to support positive and negative masking

Well this depends on what the API is, I think. In the sense that if we accept a list of intervals (or numpy equivalent) rather than a bed file, then we need only support one option, and can provide users with a utility function that inverts the intervals. If we really do what two params, I might suggest `inclusion_mask` and `exclusion_mask`.

(And now we're certainly getting into territory where we want to think hard about including support for interval operations, if only to quickly check disjointness).

> * Overloading `+` makes me very nervous.

Sorry, I hadn't intended that to be valid syntax at all. I just got sick of typing. :)

**Comment by jeromekelleher** [comment]
SGTM

**Comment by apragsdale** [comment]
That all sounds reasonable to me. A few thoughts:

If you draw random regions from across the genome of 1Mb (e.g.), you could very easily draw some "problematic" region, like the first third of chromosome 22. If you specify a mask though to avoid this issue, the keep intervals are not likely to be 1Mb long in contiguous segments (for example, the 1000G strict mask is a fairly fragmented - i.e. lots of little holes), so you will have to grab segments that are partially masked, or if you still draw randomly, some entire 1Mb segment could be masked.

I like more the idea of `mask` being the file path to some bed file instead of a list or array of intervals. As far as whether it's and inclusion vs exclusion mask, I lean toward that we should go with whatever looks to be standard format for commonly used masks. But we could support both with a single `mask` kw arg: the API could be `get_contig("chr1", ..., mask=fpath, exclude=True)`, where `exclude` is an optional argument used with `mask` that defaults to False (for an inclusion mask), and can be set to True if we have an exclusion mask. Something along those lines anyway..

Lastly, since we store recombination maps, we might want to think about storing some standard masks as well just as we do with rec maps.

**Comment by apragsdale** [comment]
Currently I opened two PRs (one for the masking, one for the generic contig) - I think I'll close one of them and combine, since all the proposed changes are related and will be easier to visualize together.

**Comment by jeromekelleher** [comment]
I suppose the mask inputs could support a path-like argument as well as the array of intervals. I'd be pretty negative about making a file path the *only* way of specifying the intervals - it just makes doing stuff programatically so much more awkward (e.g., I want to make some simple modifications to a given mask - sigh, I have to write a file to the file system just to get the API to accept it.) File paths are a very restrictive way of specifying input.

**Comment by apragsdale** [comment]
> File paths are a very restrictive way of specifying input.

I see, that's a very good point.

Passing intervals directly in the API makes sense. For the CLI, we'd want specify a path though, right? And maybe in the future the name of the mask if we store them.

To have both options, would you just check the type of the input (string vs array-like object) and go from there depending on the input type?

**Comment by jeromekelleher** [comment]
> To have both options, would you just check the type of the input (string vs array-like object) and go from there depending on the input type?

Yeah, that's what I had in mind.

**Comment by grahamgower** [comment]
> If you draw random regions from across the genome of 1Mb (e.g.), you could very easily draw some "problematic" region, like the first third of chromosome 22. If you specify a mask though to avoid this issue, the keep intervals are not likely to be 1Mb long in contiguous segments (for example, the 1000G strict mask is a fairly fragmented - i.e. lots of little holes), so you will have to grab segments that are partially masked, or if you still draw randomly, some entire 1Mb segment could be masked.

That's a really good point. I hadn't actually looked at any mask files. :(  Maybe we could use some heuristic for the proportion of the region that is allowed to be missing?

> Lastly, since we store recombination maps, we might want to think about storing some standard masks as well just as we do with rec maps.

Sounds good!

**Comment by nspope** [comment]
Closing as resolved by #1348
